src/Features/Core/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.cs

```diff
@@ -750,34 +821,59 @@ private DiagnosticData UpdatePosition(DiagnosticData diagnostic, SyntaxTree tree
                 helpLink: diagnostic.HelpLink);
         }
 
-        private static IEnumerable<DiagnosticData> GetDiagnosticData(Document document, TextSpan? span, IEnumerable<Diagnostic> diagnostics)
+        private static IEnumerable<DiagnosticData> GetDiagnosticData(Document document, SyntaxTree tree, TextSpan? span, IEnumerable<Diagnostic> diagnostics)
         {
-            return diagnostics != null ? diagnostics.Where(dx => ShouldIncludeDiagnostic(dx, span)).Select(d => DiagnosticData.Create(document, d)) : null;
+            return diagnostics != null ? diagnostics.Where(dx => ShouldIncludeDiagnostic(dx, tree, span)).Select(d => DiagnosticData.Create(document, d)) : null;
```

Comment:
AnalyzeProject only runs compilation actions, not individual symbol/node/model actions. There is multiple layer that needs re-work:

1. IDE driver when asked for diagnostics for a document, filters out all reported diagnostics which have location in tree.
2. Fixing (1) is trivial, but then we need to cache the non-local diagnostics somewhere and add it to the eventually reported project diagnostics. That part is complicated as then it adds a dependency between what all documents have had their AnalyzeDocument invoked before the solution crawler invokes AnalyzeProject.

I think the only way to solve this problem is that in v2 engine, when we are doing AnalyzeProject, just re-run all actions using CompilationWithAnalyzers and de-duplicate all reported local document diagnostics.